xmake/rules/c++/modules/xmake.lua

```diff
@@ -119,6 +119,7 @@ rule("c++.build.modules.builder")
 
                 -- cull external modules objectfile
                 compiler_support.cull_objectfiles(target, modules, sourcebatch)
+                table.sort(sourcebatch.objectfiles)
```

Comment:
This seems to solve the problem, but I feel like I should find what is actually causing the objectfiles to get out of order, maybe somewhere in the pairs call. Then use orderpairs instead of it.